### ISSUE TITLE
Remove R2 extra allocations

### ISSUE THREAD

**Issue by MohamedLaghdafHABIBOULLAH** [JuliaSmoothOptimizers/RegularizedOptimization.jl#146]
Three changes in order to make ```solve!```  of R2 non allocating.
* Define $\rho_k$ as local instead of global variable.
* Use the last version of ```ShiftedProximalOperators```in order to use non allocating ```prox!``` of L0, L1, Rank and Nuclear norm.
* Add ```test_allocs.jl``` file based on ```@wrappedallocs``` macro.
@dpo @MaxenceGollier

**Comment by MohamedLaghdafHABIBOULLAH** [comment]
Do you it's useful to precise the type of ```stats```within ```R2``` as following ?
```
stats = GenericExecutionStats(reg_nlp.model, solver_specific = Dict{Symbol, Union{Float64, Vector{Float64}, Vector{Int64}}}())
```

**Comment by dpo** [comment]
@MohamedLaghdafHABIBOULLAH Yes it definitely would, but you don’t want to use `Float64`. Instead, use the element type of the NLPModel.

**Comment by MaxenceGollier** [comment]
@MohamedLaghdafHABIBOULLAH I think you can close this issue now.
